src/khepri_fun.erl

```diff
@@ -445,22 +453,57 @@ handle_validation_error(
     {bad_type, {needed, NeededType}, {actual, any}}}},
   Error) ->
     VarInfo = {var_info, Var, [{type, NeededType}]},
-    Comment = {'%', VarInfo},
+    Comments = [{'%', VarInfo}],
     Location = {before, FailingInstruction},
-    add_comment_and_retry(Asm, Error, FailingFun, Location, Comment);
+    add_comments_and_retry(Asm, Error, FailingFun, Location, Comments);
 handle_validation_error(Asm, _ValidationFailure, Error) ->
     throw({compilation_failure, Error, Asm}).
 
-add_comment_and_retry(
-  Asm, Error, FailingFun, Location, Comment) ->
+find_comments_for_call(
+  [#function{name = Name, arity = Arity, code = Instructions} | _Rest],
+  Name, Arity, CallIndex) ->
+    find_comments_in_branch(Instructions, CallIndex);
+find_comments_for_call([_function | Rest], Name, Arity, CallIndex) ->
+    find_comments_for_call(Rest, Name, Arity, CallIndex).
+
+find_comments_in_branch(Instructions, CallIndex) ->
+    %% `beam_validator' instruction counter is 1-indexed
+    find_comments_in_branch(Instructions, CallIndex, 1, #{}).
+
+find_comments_in_branch(
+  _Instructions, CallIndex, CallIndex, VarInfos) ->
+    [{'%', {var_info, Var, Info}} || {Var, Info} <- maps:to_list(VarInfos)];
+find_comments_in_branch(
+  [{'%', {var_info, Var, Info}} | Rest], CallIndex, Counter, VarInfos) ->
+    VarInfos1 = maps:put(Var, Info, VarInfos),
+    find_comments_in_branch(Rest, CallIndex, Counter + 1, VarInfos1);
+find_comments_in_branch(
+  [{move, Src, Dst} | Rest], CallIndex, Counter, VarInfos) ->
+    VarInfos1 = case VarInfos of
+                    #{Src := Info} -> maps:put(Dst, Info, VarInfos);
+                    _ -> maps:remove(Dst, VarInfos)
```

Comment:
I usually align arrows horizontally when the clause body fits on the same line. Do you mind to do that as well?